http/src/main/java/io/micronaut/http/CaseInsensitiveMutableHttpHeaders.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.convert.ArgumentConversionContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.util.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * A {@link MutableHttpHeaders} implementation that is case-insensitive.
+ *
+ * @author Tim Yates
+ * @since 4.0.0
+ */
+@Internal
+public final class CaseInsensitiveMutableHttpHeaders implements MutableHttpHeaders {
+
+    private final boolean validate;
+    private final TreeMap<String, List<String>> backing;
+    private final ConversionService<?> conversionService;
+
+    /**
+     * Create an empty CaseInsensitiveMutableHttpHeaders.
+     *
+     * @param conversionService The conversion service
+     */
+    public CaseInsensitiveMutableHttpHeaders(ConversionService<?> conversionService) {
+        this(true, new HashMap<>(), conversionService);
```

Comment:
seems this can be `Collections.emptyMap()` since this map isn't actually used?

Reply:
Done 👍   I think my brain avoids using these due to the immutability, but we don't mutate it here so it's much better,  thanks!